### PR TITLE
The Learning Servlet

### PR DESCRIPTION
**Pull request by omniprof**
Corrected minor error in pom.xml file.
Added a new example to the Servlet Examples module called TheLearningServlet that is designed for beginners to servlets. 
Signed-off-by:Ken Fogel <omniprof@gmail.com>

### CODE REVIEW

File: servlet/theLearningServlet/pom.xml [hunk 1]
```diff
@@ -0,0 +1,86 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
Comment (manorrock) [inline servlet/theLearningServlet/pom.xml:1]: Please remove the comments from the POM.

If you want to example Maven I would suggest linking to the Maven book by Sonatype.

File: servlet/theLearningServlet/src/test/java/jakartaee/examples/servlet/thelearningservlet/TheLearningServletTest.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
+/*
```
Comment (manorrock) [inline servlet/theLearningServlet/src/test/java/jakartaee/examples/servlet/thelearningservlet/TheLearningServletTest.java:1]: Please do not use @ in JavaDoc comments unless linking to documentation that explains something

File: servlet/pom.xml [hunk 3]
```diff
@@ -19,6 +19,7 @@
         <module>pushBuilder</module>
         <module>servletContextListener</module>
         <module>servletRequestListener</module>
+        <module>theLearningServlet</module>
```
Comment (manorrock) [inline servlet/pom.xml:22]: Please rename theLearningServlet to explainingHttpServlet (as that is what you are explaining here)

**Comment by omniprof** [comment]
I have made all the requested changes. You mention the license header and I believe, based on reviewing other examples, it only appears in the Java source files and not the html, css, pom.xml or web.xml files. The new name does fit the style of the other examples.

I do object to the request to remove comments from the pom file and the html file although I have complied. If the audience to these examples are experienced developers looking to see what's new in Jakarta then maybe you can make the case that commenting these files is superfluous. I contend that many of the users of these examples are students and new developers whom we have a duty to provide with as much information about what is happening in these examples as possible.

File: servlet/explainingHttpServlet/src/main/java/jakartaee/examples/servlet/explainingHttpServlet/ExplainingHttpServlet.java [hunk 4]
```diff
@@ -10,14 +10,14 @@
...
 
 /**
- * TheLearningServlet Web Application
+ * ExplainingHttpServlet Web Application
  *
- * Prepared by Ken Fogel omniprof@gmail.com
+ * @author Ken Fogel (omniprof@gmail.com)
  *
- * Twitter: @omniprof - Blog: www.omnijava.com
+ * Twitter: omniprof - Blog: www.omnijava.com
```
Comment (manorrock) [inline servlet/explainingHttpServlet/src/main/java/jakartaee/examples/servlet/explainingHttpServlet/ExplainingHttpServlet.java:20]: Sorry I missed this one ... none of the examples add more than the @author tag so can you remove this and then I'll approve the PR and I'll ask @arjantijms Arjan to merge it.

**Comment by manorrock** [comment]
Ken your objection is noted and I do understand it, but these examples are not a guide on learning Maven, HTML or CSS. If you think we should list links to each of those respective topics in our main README I am OK with that.

File: servlet/theLearningServlet/src/main/java/jakartaee/examples/servlet/thelearningservlet/TheLearningServlet.java [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+ *
+ */
+@WebServlet(urlPatterns = "/learning")
+public class TheLearningServlet extends HttpServlet {
+
+    // All methods log their name just to show the order of calling
+    private static final Logger LOG = Logger.getLogger(TheLearningServlet.class.getName());
+
+    /**
+     * Constructor
```
Comment (pzygielo) [inline servlet/theLearningServlet/src/main/java/jakartaee/examples/servlet/thelearningservlet/TheLearningServlet.java]: I don't think this is needed. Not only it is logged in the body, but javadoc prepares separate section for constructors.

File: servlet/theLearningServlet/src/main/java/jakartaee/examples/servlet/thelearningservlet/TheLearningServlet.java [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+     *
+     * Can only be used to initialize class variables or perform tasks before
+     * the servlet comes into existence. Class variables are not thread safe so
```
Comment (pzygielo) [inline servlet/theLearningServlet/src/main/java/jakartaee/examples/servlet/thelearningservlet/TheLearningServlet.java]: It doesn't say so, but for me it suggests that in opposition - instance variables are thread safe. But are they?

File: servlet/theLearningServlet/src/main/java/jakartaee/examples/servlet/thelearningservlet/TheLearningServlet.java [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+    }
+
+    /**
+     * getter for a String of information
+     *
+     * Returns information about the servlet, such as author, version, and
+     * copyright that you can add. Usually called by the app server similar to
+     * how the attribute 'description' is used in @WebServlet.
+     *
+     * @return
```
Comment (pzygielo) [inline servlet/theLearningServlet/src/main/java/jakartaee/examples/servlet/thelearningservlet/TheLearningServlet.java]: I'd drop empty javadoc tags, as they provide no additional info, like this return above, or params/throws below.

**Comment by omniprof** [comment]
All items brought to my attention by pzygielo are addressed.

**Comment by omniprof** [comment]
I have addressed all the items you brought to my attention.

 

Thank-you,

 

Ken

**Comment by pzygielo** [comment]
> I have addressed all the items you brought to my attention. 

1. I can't see the value of empty javadoc tags, that do not provide explanations.
Unnecessary to duplicate param names or throw declarations.
Empty return tag?

It all results in:
```
...: warning: no description for @param
...: warning: no description for @throws
...: warning: no description for @return
```
from `javadoc`.

2. There is no need for FQDN in javadoc, types get resolved properly from imports.

**Comment by pzygielo** [comment]
> It all results in:
> ...

With current project configuration (i.e. implicit javadoc plugin 3.2.0), `mvn javadoc:javadoc` results in build error.

File: servlet/explainingHttpServlet/src/main/java/jakartaee/examples/servlet/explainingHttpServlet/ExplainingHttpServlet.java [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+
+    /**
+     * What follows are the four methods doGet, doPost, dotPut, and doDelete,
+     * that you can override. These also play a role in RESTful web services.
+     * There are three other methods in this family, doHead, doOptions, and
+     * doTrace, that you rarely override.
+     */
+    /**
+     * Called by the server (via the service method) to allow a servlet to
+     * handle a GET request. In a RESTful web service it is CRUD -> R
```
Comment (pzygielo) [inline servlet/explainingHttpServlet/src/main/java/jakartaee/examples/servlet/explainingHttpServlet/ExplainingHttpServlet.java:134]: javadoc doesn't like this:
```
jakartaee-examples/servlet/explainingHttpServlet/src/main/java/jakartaee/examples/servlet/explainingHttpServlet/ExplainingHttpServlet.java:171: error: bad use of '>'
     * handle a POST request. In a RESTful web service it is CRUD -> C
                                                                   ^
```

**Comment by omniprof** [comment]
Removed '->' from comments
Added info to @return in getServletInfo()
Added info to four doXXX methods describing where request and response come from.
JavaDocs shows a warning for not describing what the exceptions are for. Leaving this unchanged.

Will add creating JavaDocs to my blog and submitting to an open source project.

**Comment by arjantijms** [comment]
That was quite a process, but let's merge it in now! :)

**Comment by omniprof** [comment]
It was worth all the time I spent. Thank you to everyone who reviewed the code and went to the trouble of finding ways I could make it better.

________________________________

**Pull request merged**